logstash-core/src/main/java/org/logstash/ackedqueue/FsUtil.java

```diff
@@ -0,0 +1,47 @@
+package org.logstash.ackedqueue;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * File System Utility Methods.
+ */
+public final class FsUtil {
+
+    private FsUtil() {
+    }
+
+    /**
+     * Checks if the request number of bytes of free disk space are available under the given
+     * path.
+     * @param path Directory to check
+     * @param size Bytes of free space requested
+     * @return True iff the
+     * @throws IOException on failure to determine free space for given path's partition
+     */
+    public static boolean hasFreeSpace(final String path, final long size) throws IOException {
+        final File[] partitions = File.listRoots();
+        File location = new File(path).getCanonicalFile();
+        boolean found = false;
+        while (!found) {
+            for (final File partition : partitions) {
+                if (partition.equals(location)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                location = location.getParentFile();
+                if (location == null) {
+                    throw new IllegalStateException(
```

Comment:
I am not sure we want to throw an unchecked `IllegalStateException` here, I think a proper `IOException` would be better since reaching a `null` value seems to be a *valid* state or behaviour, per the doc:

> Returns the abstract pathname of this abstract pathname's parent, or null if this pathname does not name a parent directory.

I agree it might be a hard to test case and maybe this is OS specific too but I don't think it warrants an unchecked `IllegalStateException`. WDYT?

Reply:
I am still on the fence about this `IllegalStateException` vs `IOException` but will agree with what you think is best here.